### PR TITLE
[docs] Clean up doc structure (first part)

### PR DESCRIPTION
**Pull request by maxpumperla**
This PR addresses some of the boring ground work to further enhance and restructure our docs:

- [x] every project x is named `ray-x` consistently and every `rst` file belonging to that project is in its respective subfolder. This was mostly an issue with `core` and `rllib`. This makes the overall structure a bit less intimidating, especially for potentially new contributors.
- [x] all cloud scripts are consolidated in a `tools` folder.
- [x] We now have `dev` dependencies to allow for faster build times. First build is the usual ~2mins, rebuilds can go down to about 8 seconds now, which is massive for productivity.
- [x] The README reflects the changes and also shows how to only build individual subprojects, which reduces build time to around 4 secs in most cases.
- [x] There's a common `_includes` folder now for better alignment of announcements (was very scattered before)
- [x] The big `examples` folder, which was _mostly_ about Ray core (and located outside of `sources`, has been migrated into `ray-core` for alignment with other libraries. The examples that don't belong to core should later be moved to the right places (not in this PR).
- [x] Doctests have been activated and can now be run with `make doctest`. There are currently ~500 failing doc tests. I fixed one to show how, but this needs to be a follow-up engineering-wide effort to fix properly. (Having doctest as part of CI would be another follow-up.
- [x] We're using `sphinx-external-toc` to have an easier to use yaml file of the TOC. We should avoid hidden, partial TOCs within docs. No easy way to "get" the structure that way.

Due to the nature of this PR, you might want to cherry-pick individual commits for review. I've tried to keep commits clean and have the renaming/moving/decluttering parts lumped into `1a2e8c1`. That commit is so large that it doesn't render the diff on GitHub.

### CODE REVIEW

**Comment by maxpumperla** [comment]
@richardliaw I took the best of your WIP PR and made it work in a `dev` setting, as documented.

I'm not sure what the merge conflict above is supposed to be (I've moved that file elsewhere), but I don't have the rights to resolve it. :/

**Comment by richardliaw** [comment]
> every project x is named ray-x consistently and every rst file belonging to that project is in their respective subfolder. This was mostly an issue with core and rllib. This makes the overall structure a bit less intimidating, especially for potentially new contributors.

Do we expect to break all external links this way?

File: doc/requirements-rtd.txt
Comment (richardliaw) [inline doc/requirements-rtd.txt:1]: The reason why this is here / in the instructions is because it allows us to reproduce the sphinx build on readthedocs exactly. Is there a reason to remove?

File: doc/source/conf.py
Comment (richardliaw) [inline doc/source/conf.py:266]: can you autopopulate this?

File: doc/source/conf.py
Comment (richardliaw) [inline doc/source/conf.py:160]: hmm will this make it so that all 'ray.experimental' objects don't get rendered properly?

File: doc/source/_includes/rllib_we_are_hiring.rst
Comment (richardliaw) [inline doc/source/_includes/rllib_we_are_hiring.rst:1]: i think there's another hiring post (not just rllib) somewhere?

File: rllib/offline/mixed_input.py
Comment (richardliaw) [inline rllib/offline/mixed_input.py:17]: nice..........

File: doc/requirements-rtd.txt
Comment (maxpumperla) [reply doc/requirements-rtd.txt:1]: accidental, thanks for spotting this.

File: doc/source/_includes/rllib_we_are_hiring.rst
Comment (maxpumperla) [reply doc/source/_includes/rllib_we_are_hiring.rst:1]: yes, in the "cluster" part.

File: doc/source/conf.py
Comment (maxpumperla) [reply doc/source/conf.py]: done

File: doc/source/conf.py
Comment (maxpumperla) [reply doc/source/conf.py]: yes, this was just one occurrence though. I've removed it from docs, as it didn't add any value. This way we don't need the above hack.

File: doc/source/_includes/rllib_we_are_hiring.rst
Comment (maxpumperla) [reply doc/source/_includes/rllib_we_are_hiring.rst]: to be clear, I grouped all messaging into one folder now so that we can consolidate it in follow-up PRs. the "rllib" hiring thingy is specific to rllib right now, but should it be? maybe. we could establish some best practices here. Also, Asawari and co. might find this setup easier to announce Ray summits etc. more easily.

File: doc/requirements-dev.txt
Comment (richardliaw) [inline doc/requirements-dev.txt:29]: pin?

**Review by richardliaw — APPROVED** [review]
Looks good to me!

File: doc/requirements-dev.txt
Comment (maxpumperla) [reply doc/requirements-dev.txt]: done

**Comment by maxpumperla** [comment]
@ericl @scv119 @iycheng could you please help stamp this? We need code owner approval.

**Comment by maxpumperla** [comment]
> > every project x is named ray-x consistently and every rst file belonging to that project is in their respective subfolder. This was mostly an issue with core and rllib. This makes the overall structure a bit less intimidating, especially for potentially new contributors.
> 
> Do we expect to break all external links this way?

yes, as discussed, to change the structure at all, we have to do it at some point. and we can set up redirect rules for 404s in readthedocs to catch this. :v:

**Review by ericl — APPROVED** [review]

**Comment by zhe-thoughts** [comment]
Terrific work @maxpumperla ! Not sure how easy it is:

Is it possible to describe the new layout? Something like:
`doc/source`
\- `data`
\- - ...
\- Other libraries
\- Generic

Basically, I'm starting to work on a wiki page telling developers how to add documentation when they add code, which requires telling them the structure

**Comment by simon-mo** [comment]
@maxpumperla I think this makes all the old link broken, which is linked in various places including stackoverflow, discuss, slack, google search, etc. 

Is there a way to preserve/duplicate the url. For example 

works: https://docs.ray.io/en/latest/serve/index.html
broke: https://docs.ray.io/en/master/serve/index.html

**Comment by maxpumperla** [comment]
@simon-mo we have to define redirect rules in readthedocs.org. 

usually better not to link to latest (I realise the trade-offs)

**Comment by simon-mo** [comment]
Is there a way to leave the old entries but put in redirect link in the page content? (like don't put it in the toc index, or hide them somehow).

**Comment by simon-mo** [comment]
> usually better not to link to latest (I realise the trade-offs)

We might need docs.ray.io redirect to latest release version instead of /latest.

**Comment by xwjiang2010** [comment]
btw, next time we merge it, we may want to change this path: `NOTEBOOK_PATH_RELATIVE_TO_RAY_REPO`

https://buildkite.com/ray-project/periodic-ci/builds/2437#ed02277e-5d27-4bff-a88e-d842f30e3f28

Some release tests are relying on this hard coded notebook path :)

**Pull request merged**
